docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+* It shall become much easier to move control planes of Shoot clusters between Seed clusters ([#232](https://github.com/gardener/gardener/issues/232)) which is a necessary requirement of an automated setup for the Gardener Ring ([#233](https://github.com/gardener/gardener/issues/233)).
+
+### Non-Goals
+
+* We want to also factor out the specific knowledge of the addon deployments (nginx-ingress, kubernetes-dashboard, ...), but we already have dedicated projects/issues for that: https://github.com/gardener/bouquet and [#246](https://github.com/gardener/gardener/issues/246). We will keep the addons in-tree as part of this proposal and tackle their extraction separately.
+* We do not want to make the Gardener a plain workflow engine that just executes a given template (which indeed would allow to be generic, open, and extensible in their highest forms but which would end-up in building a "programming/scripting language" inside a serialization format (YAML/JSON/...)). Rather, we want to have well-defined contracts and APIs.
+
+## Proposal
+
+Gardener heavily relies on and implements Kubernetes principles, and its first goal is to use Kubernetes wherever applicable.
+The extension concept in Kubernetes embraces (next to others) `CustomResourceDefinition`s as well as `ValidatingWebhookConfiguration`s and `MutatingWebhookConfiguration`s.
+Consequently, Gardener's extensibility concept relies on these mechanisms.
+
+Instead of doing everything on its own, Gardener will deploy a bunch of CRDs to the Seed cluster which will be watched by dedicated controllers only containing the specific knowledge of their function area.
+Gardener will just wait until the controllers report that they are done (or have faced an error) in the CRD's `.status` field instead of doing the respective tasks itself.
+We will have one specific CRD for every specific operation (e.g., DNS, infrastructure provisioning, machine cloud config generation, ...).
+However, there are also parts inside Gardener which are handled generally (not by cloud botanists) because they are the same or very similar for all the environments.
+One example of those is the deployment of a `Namespace` in the Seed which will run the Shoot's control plane
+Another one is the deployment of a `Service` for the Shoot's kube-apiserver.
+In case a cloud botanist needs to interfere and react on those operations it should register a `ValidatingWebhookConfiguration` or a `MutatingWebhookConfiguration`.
+With this approach it can validate or modify any resource created by Gardener to make it cloud specific.
+
+### Modification of existing `CloudProfile` and `Shoot` resources
+
+We will introduce the new API group `gardener.cloud`:
+
+#### CloudProfiles
+
+```yaml
+---
+apiVersion: gardener.cloud/v1alpha1
+kind: CloudProfile
+metadata:
+  name: aws
+spec:
```

Comment:
One way could be omitting the spec in gardener and let cloud provider controllers place some validating webhooks as suggested in
https://github.com/gardener/gardener/pull/370/files#diff-211ac3b72ed2dbadb74b3e8304eb4c71R66

Reply:
We could omit the spec, but would you put every key on top level or just rename it?

Reply:
I'm not sure if this comment is for me. If so I didnt really meant to omit the spec, my comment is to address how to execute the validation of the cloud provider specific `providerConfig` fields since it's contents are meant to be defined out of tree controllers.

Reply:
@bergerx I'll add a section about that, thanks for pointing that out.